resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -478,15 +479,36 @@ public DnsNameResolver(
             @Override
             protected void initChannel(DatagramChannel ch) {
                 ch.pipeline().addLast(DATAGRAM_ENCODER, DATAGRAM_DECODER, responseHandler);
+                ch.closeFuture().addListener((ChannelFutureListener) future1 -> {
+                    resolveCache.clear();
+                    cnameCache.clear();
+                    authoritativeDnsServerCache.clear();
+                });
             }
         });
+        b.option(ChannelOption.RCVBUF_ALLOCATOR, new FixedRecvByteBufAllocator(maxPayloadSize));
 
         channelFuture = responseHandler.channelActivePromise;
-        final ChannelFuture future;
+        Future<Channel> channelFuture = b.createUnregistered();
+        if (!channelFuture.isSuccess()) {
+            Throwable cause = channelFuture.cause();
+            if (cause != null) {
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                }
+                if (cause instanceof Error) {
+                    throw (Error) cause;
+                }
+            }
+            throw new IllegalStateException(
```

Comment:
This looks hard to understand as to why we expect the future to complete immediately. I understand why it is done this way to make sure that we can initialize the `final` variable `ch`. Should we make it non-final and fix the other code appropriately? OR maybe like what @ejona86 is suggesting in https://github.com/netty/netty/pull/11517#issuecomment-888455321; we just have `createUnregistered()` return a `Channel`

Reply:
I like the idea of @ejona86